File: src/org/cerion/symcalc/expression/number/RealNum_Double.java
package org.cerion.symcalc.expression.number;

import org.jetbrains.annotations.NotNull;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

class RealNum_Double extends RealNum {

    private double dNumber = 0;

    private RealNum_Double() {
        this.dNumber = 0;
    }

    RealNum_Double(String s) {
        dNumber = Double.parseDouble(s);
    }

    RealNum_Double(IntegerNum n) {
        dNumber = n.toDouble();
    }

    RealNum_Double(RationalNum r) {
        dNumber = r.toDouble();
    }

    RealNum_Double(double n) {
        dNumber = n;
    }

    @Override
    public double toDouble() {
        return dNumber;
    }

    @Override
    public String toString() {
        return "" + dNumber;
    }

    @Override
    public boolean equals(NumberExpr e) {
        if(e.isReal()) {
            if (e instanceof RealNum_BigDecimal)
                throw new NotImplementedException();

            return this.dNumber == e.toDouble();
        }

        return false;
    }

    public RealNum_Double unaryMinus() {
        return new RealNum_Double(0 - dNumber);
    }

    public boolean isZero() {
        return (dNumber == 0.0);
    }

    public boolean isOne() {
        return (dNumber == 1.0);
    }

    public NumberExpr plus(NumberExpr num) {
        RealNum_Double result = new RealNum_Double();
        switch (num.getNumType())
        {
            case INTEGER: //RealNum + IntegerNum
                result.dNumber = dNumber + num.toDouble();
                return result;

            case RATIONAL: //RealNum + RationalNum
            {
                result.dNumber = dNumber + num.toDouble();
                return result;
            }

            case REAL: //RealNum + RealNum
            {
                //result.bigNumber = this.bigNumber.add( ((RealNum)num).bigNumber );
                result.dNumber = dNumber + ((RealNum_Double)num).dNumber;
                return result;
            }

        }

        return num.plus(this);
    }

    public NumberExpr minus(NumberExpr num) {
        RealNum_Double result = new RealNum_Double();
        switch (num.getNumType())
        {
            case INTEGER: //RealNum - IntegerNum
            {
                result.dNumber = dNumber - num.toDouble();
                return result;
            }

            case RATIONAL: //RealNum - RationalNum
            {
                result.dNumber = dNumber - num.toDouble();
                return result;
            }

            case REAL: //RealNum - RealNum
            {
                result.dNumber = dNumber - ((RealNum_Double)num).dNumber;
                return result;
            }
        }

        NumberExpr negative = num.unaryMinus();
        return negative.plus(this);
    }

    public NumberExpr times(NumberExpr num) {
        RealNum_Double result = new RealNum_Double();
        switch (num.getNumType())
        {
            case INTEGER: //RealNum * IntegerNum
                result.dNumber = dNumber * num.toDouble();
                return result;

            case RATIONAL: //RealNum * RationalNum
            {
                result.dNumber = dNumber * num.toDouble();
                return result;
            }

            case REAL: //RealNum * RealNum
            {
                result.dNumber = dNumber * ((RealNum_Double)num).dNumber;
                return result;
            }
        }
        return num.times(this);
    }

    public NumberExpr div(NumberExpr num) {
        RealNum_Double result = new RealNum_Double();
        switch (num.getNumType())
        {
            case INTEGER: //RealNum / IntegerNum
                result.dNumber = dNumber / num.toDouble();
                return result;

            case RATIONAL: //RealNum / RationalNum
            {
                result.dNumber = dNumber / num.toDouble();
                return result;
            }

            case REAL: //RealNum / RealNum
            {
                result.dNumber = dNumber / ((RealNum_Double)num).dNumber;
                return result;
            }
        }
        return num.times(this);
    }

    public NumberExpr power(NumberExpr num) {
        RealNum_Double result = new RealNum_Double();
        switch (num.getNumType())
        {
            case INTEGER: //RealNum ^ IntegerNum
            case RATIONAL: //RealNum ^ RationalNum
            case REAL: //RealNum ^ RealNum
            {
                result.dNumber = Math.pow( dNumber, num.toDouble());
                return result;
            }
        }

        return null;
    }

    @Override
    public boolean isWholeNumber() {
        if(dNumber == Math.floor(dNumber) && !Double.isInfinite(dNumber))
            return true;

        return false;
    }

    @Override
    public IntegerNum toInteger() {
        return new IntegerNum((long)dNumber);
    }

    @NotNull
    @Override
    public NumberType getNumType() {
        return NumberType.REAL;
    }
}

File: src/expression/number/RealNum_Double.kt
package org.cerion.symcalc.expression.number

import kotlin.math.floor

class RealNum_Double(override val value: Double = 0.0) : NumberExpr() {

    override val numType: NumberType get() = NumberType.REAL_DOUBLE
    override val isZero: Boolean get() = value == 0.0
    override val isOne: Boolean get() = value == 1.0
    override val isNegative: Boolean get() = value < 0
    override val precision: Int get() = SYSTEM_DECIMAL_PRECISION

    val isWholeNumber: Boolean
        get() = value == floor(value) && !java.lang.Double.isInfinite(value)

    override fun toString(): String = "" + value

    override fun evaluate(precision: Int): NumberExpr {
        return this // Cannot upgrade precision so do nothing
    }

    override fun compareTo(other: NumberExpr): Int {
        when(other.numType) {
            NumberType.INTEGER,
            NumberType.RATIONAL -> return this.compareTo(other.evaluate(SYSTEM_DECIMAL_PRECISION))
            NumberType.REAL_DOUBLE -> return value.compareTo(other.asDouble().value)
            NumberType.REAL_BIGDEC -> return value.compareTo(other.asBigDec().toDouble())
            NumberType.COMPLEX -> return Complex(this).compareTo(other)

        }
    }

    override fun unaryMinus(): RealNum_Double = RealNum_Double(0 - value)

    override fun plus(other: NumberExpr): NumberExpr {
        when (other.numType) {
            NumberType.INTEGER,
            NumberType.RATIONAL -> return this + other.evaluate(SYSTEM_DECIMAL_PRECISION)
            NumberType.REAL_DOUBLE -> return RealNum_Double(value + other.asDouble().value)
            NumberType.REAL_BIGDEC -> return RealNum_Double(value + other.asBigDec().toDouble())
            NumberType.COMPLEX -> return Complex(this) + other
        }
    }

    override fun times(other: NumberExpr): NumberExpr {
        when (other.numType) {
            NumberType.INTEGER,
            NumberType.RATIONAL -> return this * other.evaluate(SYSTEM_DECIMAL_PRECISION)
            NumberType.REAL_DOUBLE -> return RealNum_Double(value * other.asDouble().value)
            NumberType.REAL_BIGDEC -> return RealNum_Double(value * other.asBigDec().toDouble())
            NumberType.COMPLEX -> return Complex(this) * other

        }
    }

    override fun div(other: NumberExpr): NumberExpr {
        when (other.numType) {
            NumberType.INTEGER,
            NumberType.RATIONAL -> return this / other.evaluate(SYSTEM_DECIMAL_PRECISION)
            NumberType.REAL_DOUBLE -> return RealNum_Double(value / other.asDouble().value)
            NumberType.REAL_BIGDEC -> return RealNum_Double(value / other.asBigDec().toDouble())
            NumberType.COMPLEX -> return Complex(this) / other
        }
    }
}
